candle-core/benches/bench_utils.rs

```diff
@@ -0,0 +1,56 @@
+use candle_core::{Device, Result};
+
+pub(crate) trait BenchDevice {
+    fn sync(&self) -> Result<()>;
+}
+
+impl BenchDevice for Device {
+    fn sync(&self) -> Result<()> {
+        match self {
+            Device::Cpu => Ok(()),
+            Device::Cuda(device) => {
+                #[cfg(feature = "cuda")]
+                return Ok(device.synchronize()?);
+                #[cfg(not(feature = "cuda"))]
+                panic!("Cuda device without cuda feature enabled: {:?}", device)
+            }
+            Device::Metal(device) => {
+                #[cfg(feature = "metal")]
+                return Ok(device.wait_until_completed()?);
+                #[cfg(not(feature = "metal"))]
+                panic!("Metal device without metal feature enabled: {:?}", device)
+            }
+        }
+    }
+}
+
+#[allow(dead_code)]
```

Comment:
Do we really need this allow pragma? Maybe there is a way to avoid it.

Reply:
I agree, but I don't know a way to get around it. Criterion doesn't run the code in the same way as the standard rust testing framework, so it doesn't register as being used.

Reply:
Ok I guess you googled a bit on the criterion side and haven't found anything (I haven't checked), could you add a comment to explain why these are needed? (and link to any criterion resource that would document this if any)

Reply:
Ok, cargo isn't complaining about unused code anymore 👍

Reply:
Are you sure? I just tried running the benchmarks with no feature enabled and got the warnings?
<img width="414" alt="image" src="https://github.com/huggingface/candle/assets/1041292/b8629685-6859-4ff3-8899-112fee252edc">

Reply:
Hmm. I tried after cargo clean and still get no warnings 🤔

Reply:
I think this is related to the `/benches` folder file structure https://users.rust-lang.org/t/how-to-properly-add-test-cases-for-benches/102859/4

Reply:
Are the warnings still present? If they are still there for any of you I'll try a different approach.

Reply:
Yes. Just remove the utils file. Bench will walk all files including the utils which is why it will complain. Just but them in the core file.

Reply:
By "core" file, do you mean `matmul.rs`?
I was thinking we could have several benchmarks `matmul.rs`, `random.rs`, etc.
If I place these in their own module and load them from a `bench_main.rs` file we won't be able to specify which benchmark to run.